### DISCUSSION TITLE
Detecting when playback has finished to play specified tracks one after another

### DISCUSSION THREAD

**Discussion by ashleycox** [Makuna/DFMiniMp3#73] [Q&A] [answered]
Hoping someone can help. I’m trying to play files one after the other by specifying the file numbers, but I can’t figure out how to have the loop wait for the current file to finish before playing the next file. The files are speech samples, used to add text-to-speech functionality to an Arduino project. I want the files to play back to back with no (or as small as possible) delay as if they were joined together. So basically:

object.play(1);
As soon as  1 has finished:
object.Play(2);
As soon as 2 has finished:
object.play(3);
And so on.

The part I’m stuck on is getting the files of different lengths to play in this way, without resorting to delays and knowing the exact length. The files always get cut off. I’m basically struggling with the ‘if is playing’ bit. I've tried using the module's 'busy' pin with digitalRead but I'm still not getting anywhere. I'm sure there's something simple I'm missing, but I'd be grateful if someone can point me in the right direction or even offer a code sample to achieve what I'm trying to do. The DFPlayer mini is connected to an Arduino Nano using software serial. Any suggestions greatly appreciated.

**Comment by Makuna** [comment, marked as answer]
In your notification class (examples use the name Mp3Notify), the method `OnPlayFinished(DfMp3& mp3, DfMp3_PlaySources source, uint16_t track)` will be called when any track finishes playing.  Just implement code that uses the passed track to figure out what the next track should be and call play with it.

**Reply by ashleycox** [reply]
Thanks for your reply. I've tried using the notification class but I still can't figure it out. Can you provide a short example?

**Reply by Schallbert** [reply]
Hi,

I used the notification class to let my application know when a track finishes,  example see here: https://github.com/Schallbert/Tonuino_alternativeFirmware/blob/ff0174a62088e3d042e5bb78c4bf30fbc8a7a6ca/lib/Arduino/DFMiniMp3_implementation.h#L175

I'm using this for prompts only. 
To play tracks consecutively, I'm simply using the player's digital play status pin: https://github.com/Schallbert/Tonuino_alternativeFirmware/blob/ff0174a62088e3d042e5bb78c4bf30fbc8a7a6ca/lib/Mp3/Mp3Play/Mp3Play_implementation.cpp#L67

Hope this helps.

**Reply by terrypin999** [reply]
As a non-programmer I find this library (and indeed GitHub itself) dauntingly complicated!

Of course, I suppose Makuna developed it with experienced developers in mind, not hobbyist end users like me. I eventually made progress with the original, "DFRobotDFPlayerMini.h" by Angelo qiao,and have recently discovered PowerBroker’s promising DFMiniMp3 which I’m about to work with. But I so do wish that all three (and future authors)  went the extra mile with more comprehensive and heavily commented examples. Such as:
Play folder \06 and stop
Play folder \23 and repeat it (loop)
Play each track for a specified duration, then delay say 2 s and play the next
Play each track for a specified duration, then fade out over say 3s and play the next
Play each track entirely, then delay say 2 s and play the next
Switch immediately to playing the folder (or all) in random order
Ditto, but wait until current track ends before switching
Jump forward or back say 10 tracks
On power-up optionally start on a randomly chosen track
etc

I initially discovered this attractive SD module a couple of months ago. But gave up because I hit  so many coding obstacles. I tried DFRobot’s USB version (‘Pro’) instead, using the library ‘DFRobot_DF1201S’. And eventually made the music box shown below as a Christmas present for my wife. But its major downside of 128 MB means that even randomised tracks are becoming familiar! So I’m now determined to replace it with the SD module. Providing I can retain the panel and its buttons, the fruits of two week’s work! 

![20211227-112204](https://user-images.githubusercontent.com/7612034/150637317-07c338b0-cc40-40e6-9408-8273d9a200fb.jpg)

**Reply by ashleycox** [reply]
Most of what you want would be done using simple conditional logic in your program, rather than in the library itself. There's an MP3 player example on my GitHub that should give you some examples of the code you need, but reach out to me and I'd be happy to write you some quick examples for the specifics above.

**Reply by terrypin999** [reply]
Thanks Ashley, much appreciate the fast and helpful reply. (I was just coming back to replace my link with an embedded image. which I'll do next.)

That's a very kind offer of yours and I'm pretty sure I'll need to take you up on it! First I'll see how far I get myself, although I suspect that will be with one of the two earlier libraries I mentioned, probably PowerBroker's latest. At this stage on the learning curve I am unashamedly frequently in copy/paste mode. But I far prefer to actually **understand** the code I'm using ;-)
Terry, UK

**Reply by Makuna** [reply]
@terrypin999 I did target Arduino customers, but "notifications" concepts in general are not a simple topic and this device and the idea that you want to know when a song completes (a notification) requires a better knowledge of programming than copy and paste to "understand how it works".  
Feel free to ask questions here or the gitter channel (https://gitter.im/DFMiniMp3/Lobby).  
Asking on gitter is more ephemeral and some find it a more comfortable place to open up and ask questions they may think are stupid. Asking here is more searchable and retained.  
I try to summarize interesting topics into the Wiki and its FAQ section (https://github.com/Makuna/DFMiniMp3/wiki) but it takes time and help is always appreciated.

**Reply by terrypin999** [reply]
Thanks Michael. I spent a few hours today returning to the original DFRobot library but once again it failed to reliably implement the sort of essential operations that I listed earlier. I’m now  intending to work with PowerBroker’s library, but will later explore your more complex approach.

BTW, the  FAQ you mention seemed to be empty, reporting ‘Under Construction’.

**Reply by jeremiawuzza** [reply]
I too have this problem (which one?) no seriously, If in my loop I call (say)
dfmp3.playMp3FolderTrack(3);
It starts to play, but if there is no delay() in the loop it only plays a short burst then repeats.
I put a delay() in the loop:
If the delay time is shorter than the track, the track is truncated - if it is longer there is a pause before the track repeats 
Another track may be a different length and so it can't be controlled by different delay times. Or at least not precisely.

I see the
OnPlayFinished(DfMp3& mp3, DfMp3_PlaySources source, uint16_t track) method but I am not sure how to use it
 for an sd card and say track 3 would it be
OnPlayFinished(DfMp3& mp3, DfMp3_PlaySources sd, 3)  ?
And where does this go - in the loop somehow? or does it call another routine?

Please help - I am at the threshold of creating a beautiful thing with your <DFMiniMp3.h>!

**Reply by Makuna** [reply]
@jeremiawuzza Did you look at the examples?  PlayMp3 example, demonstrates "implementing" OnPlayFinished().  It is not something you call, it's something you implement and that will get called when any track is finished.  
The [Wiki](https://github.com/Makuna/DFMiniMp3/wiki/Notification-Method#static-void-onplayfinisheddfmp3_playsources-source-uint16_t-track) entry for it describes what the arguments will be when called and the examples

**Reply by Makuna** [reply]
I just updated the PlayMp3 example to play tracks 1-3 sequentially, looping forever.  This demonstrates the simple technique.
You will have to pull (git term) of the source until the next release is made.
